content/2019-02-19_osmnacty-sraz-git.md

```diff
@@ -0,0 +1,71 @@
+
+Jako první věc nám Martin ukázal, že si git můžeme přizpůsobit "na míru". Příkaz `git config --global` známe z [instalace gitu](https://naucse.python.cz/course/pyladies/git/install/), kde jsme jím nastavovaly uživatelské jméno a email. Pokud ale zadáme `git config --global -e`, dostaneme se do editovatelného nastavení, kde toho jde udělat mnohem více, například nastavit aliasy pro různé přikazy, které v gitu používáme. Můžeme si tak definovat vlastní zkratky i pro delší sekvence, nejen pro jednoslovné příkazy. 
+
+![aliasy](./images/alias.png)
+
+Dostaly jsme ještě nějaké tipy k příkazové řádce obecně. Používáme pro pohyb v historii příkazů šipku nahoru, ale pomocí kombinace kláves Ctrl + R v ní mohu přímo vyhedávat zadáním požadovaného výrazu, najde jeho poslední výskyt. Pro zobrazení celé historie stačí zadat příkaz `history`.
+
+### git stash
+
+Poté, co jsme si pohrály s nastavením, posunuly jsme se k novým příkazům. Jako první přišel na řadu `git stash`. Tento příkaz může být užitečný například ve chvíli, kdy máme rozdělanou nějakou práci, ale náhle potřebujeme udělat něco jiného. Nechceme ještě dělat commit, ale můžeme využít stash, který uloží změny mimo větev, pouze lokálně. Stashů můžeme udělat tolik, kolik potřebujeme.
+Stash se zobrazuje s pořadovým číslem, podle kterého je identifikovatelný: `stash@{0}`označuje nejnověji uložený, `stash@{1}` druhý nejnovější, atd. Pokud mu nepřidáme žádný popis, přiřadí se automaticky kód, který se vypíše za pořadové číslo. Pokud mu popis dáme, vypíše se místo kódu. 
+Související příkazy:
+`git stash list` vypíše seznam všech stashů
+`git stash push -m "název"` dá stashi popis
+`git stash pop` vytahuje věci ze stashe od nejnovějšího záznamu. Pokud za tento příkaz přidám pořadové číslo stashe, například `git stash pop stash@{3}`, vytáhne ten konkrétní stash.
+`git stash drop` maže věci ze stashe od nejnovějšího záznamu, ale podobně jako pop umí mazat i konkrétní stash, pokud mu dám její pořadové číslo.
+`git stash clear` kompletně vyčistí stash, smaže všechny uložené stashe
+Git takto pracuje pouze se soubory, které zná, které už do něj byly přidány. Pokud máme nějaké, které ještě do gitu přidány nebyly a git je netrackuje, ale přesto je chceme dát do stashe, je potřeba zadat `git stash push --include-untracked`.
+Výhoda i nevýhoda stashe spočívá právě v tom, že ukládá práci mimo větve, visí volně v prostoru. Můžeme pracovat v jedné větvi, dát věci do stashe, přesunout se do nové větve a tam si změny vytáhnout. V tom ale spočívá i slabina - o data lze snadno přijít. Jakmile zavřeme příkazovou řádku, stash se smaže. Jde v podstatě jen o dočasné úložiště rozpracovaných vecí.
```

Comment:
> Jakmile zavřeme příkazovou řádku, stash se smaže.

Je to pravda? Zkusila jsem si to ted, a stash zustal na miste i po zapnuti noveho terminalu.

Reply:
Nevím, ověřovala jsem si jen věci, kterým jsem nerozuměla. Tohle zaznělo několikrát, ale zkusím to dohledat v dokumentaci.

Reply:
Hmmm zajímavé, ani dokumentace, ani různé návody se o tom nezmiňují, ale přitom to by právě podle Martina mohl být ten faktor, proč je třeba s tím pracovat opatrně. 
Pomůže @encukou?

Reply:
stash přežije i vypnutí počítače. Viz můj komentář níže k tomu samému odstavci. https://github.com/PyLadiesCZ/roboprojekt-blog/pull/37/files#r261861924

Reply:
Tak já nevím, tak v čem to je teda tak problematické?

Reply:
Z textu smažu, dík moc

Reply:
Pokud to člověk nepoužívá s rozmyslem, tak pak máš snadno v repozitáři 10 stashů a už se ti pak nechce dohledávat, co v nich skutečně máš... To je ale spíš už záležitost.. _morálky_ se mi chce říct. Velký úklid (=smaž všechny stashe) v případě potřeby je snadný (`git stash clear`).

Reply:
Aha! Dík moc :)